datafusion-federation/src/sql/schema.rs

```diff
@@ -1,45 +1,68 @@
-use std::{any::Any, sync::Arc};
+use std::{any::Any, str::FromStr, sync::Arc};
 
 use async_trait::async_trait;
-use datafusion::logical_expr::{TableSource, TableType};
-use datafusion::{
-    arrow::datatypes::SchemaRef, catalog::SchemaProvider, datasource::TableProvider, error::Result,
-};
+use datafusion::{catalog::SchemaProvider, datasource::TableProvider, error::Result};
 use futures::future::join_all;
 
-use crate::{
-    sql::SQLFederationProvider, FederatedTableProviderAdaptor, FederatedTableSource,
-    FederationProvider,
+use super::{
+    table::{DynSqlTable, SQLTable},
+    RemoteTableRef, SQLTableSource,
 };
+use crate::{sql::SQLFederationProvider, FederatedTableProviderAdaptor};
 
 #[derive(Debug)]
 pub struct SQLSchemaProvider {
-    // provider: Arc<SQLFederationProvider>,
-    tables: Vec<Arc<SQLTableSource>>,
+    tables: Vec<DynSqlTable>,
 }
 
 impl SQLSchemaProvider {
     pub async fn new(provider: Arc<SQLFederationProvider>) -> Result<Self> {
-        let tables = Arc::clone(&provider).executor.table_names().await?;
+        let executor = Arc::clone(&provider.executor);
+        let tables = executor.table_names().await?;
+        let tasks = tables
+            .iter()
+            .map(|table_name| {
+                let provider = Arc::clone(&provider);
+                let table_name = table_name.clone();
+                async move {
+                    let table_ref = RemoteTableRef::from_str(&table_name)?;
+                    let table_source = SQLTableSource::new(provider, table_ref).await;
+                    table_source.map(|table| Arc::new(table) as Arc<dyn SQLTable>)
+                }
+            })
+            .collect::<Vec<_>>();
+        let tables = join_all(tasks)
+            .await
+            .into_iter()
+            .collect::<Result<Vec<_>>>()?;
 
-        Self::new_with_tables(provider, tables).await
+        Ok(Self::new_with_tables(tables))
     }
 
-    pub async fn new_with_tables(
+    pub async fn new_with_table_names(
```

Comment:
If there is a need to change this API signature, please explain the motive in detail in the PR. Same for sql/table.rs.

Reply:
Please take a look, tried to keep the API similar